### DISCUSSION TITLE
Trouble adding a store.

### DISCUSSION THREAD

**Discussion by rmp5s** [jef/streetmerchant#1224] [Help] [answered]
I followed the instructions [here](https://github.com/jef/streetmerchant/wiki/Help:-Configuration:-Adding-a-store), and viewed the example [here](https://github.com/jef/streetmerchant/commit/af96c5f2e808af7496f3c3299e4cf173105de48b), but I keep getting errors like this:

```
src/store/lookup.ts(191,4): error TS7053: Element implicitly has an 'any' type because expression of type 'Series' can't be used to index type '{ 3070: number; 3080: number; 3090: number; rx6800: number; rx6800xt: number; rx6900xt: number; ryzen5600: number; ryzen5800: number; ryzen5900: number; ryzen5950: number; sf: number; sonyps5c: number; sonyps5de: number; 'test:series': number; xboxss: number; xboxsx: number; }'.
  Property 'throttles' does not exist on type '{ 3070: number; 3080: number; 3090: number; rx6800: number; rx6800xt: number; rx6900xt: number; ryzen5600: number; ryzen5800: number; ryzen5900: number; ryzen5950: number; sf: number; sonyps5c: number; sonyps5de: number; 'test:series': number; xboxss: number; xboxsx: number; }'.
```

I'm unsure what I'm doing wrong.  So unsure that I can't even form a very well thought out question or provide info that I think will be useful.  I don't even know where to start.  (Yes, I'm TOTALLY new to Node.JS...)

I DO think this will be useful:

[Store I'm trying to add.](https://flyhoneycomb.com/)
Items I'm trying to add - [1](https://flyhoneycomb.com/products/alpha-flight-controls), https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/bravo-throttle-quadrant, https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/airbus-throttle-handles

Happy to provide more info if you need.  Just let me know.

Thank you!

**Comment by attaway** [comment]
having the same problem

**Comment by jef** [comment]
Do you mind posting the code you're trying to add? Specifically the new store? I can help!

**Comment by rmp5s** [comment]
> Do you mind posting the code you're trying to add? Specifically the new store? I can help!

Absolutely!  Thanks for writing/releasing such an awesome utility, too.  :-D

```typescript
import {Store} from './store';

export const Honeycomb: Store = {
	labels: {
		inStock: {
			container: '#product_form_4533457682541',
			text: ['Add to cart']
		},
	},
	links: [
		{
			brand: 'test:brand',
			model: 'test:model',
			series: 'test:series',
			url: 'https://flyhoneycomb.com'
		},
		{
			brand: 'honeycomb',
			model: 'AlphaFlightYoke',
			series: 'yoke',
			url: 'https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/alpha-flight-controls'
		},
		{
			brand: 'honeycomb',
			model: 'BravoThrottleQuadrant',
			series: 'quadrant',
			url: 'https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/bravo-throttle-quadrant'
		},
		{
			brand: 'honeycomb',
			model: 'ThrottlePackForAirbus',
			series: 'throttles',
			url: 'https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/airbus-throttle-handles'
		},
	],
	name: 'honeycomb'
};
```

**Comment by Kpa62** [comment]
I drove myself crazy over this one for a longtime but figured out how to debug. I believe there is some confusion here. I'm not a coder at all but from fiddling around I ran into this. This current error is not a store error but a product error. You need to add the model, series and brand to the store.TS file. Then you will most likely get a max price error next, so then you have to add to the config.ts the max price series. also update the .env file for this series and model in the max price section. Also update the Readme.md file. finally it should all run.

What confusion I see here though is that you're adding a store before adding a model or series for the product. you need to build both. A store and a product as the flight stick is not a tracked product by default. 

I ran into this trying to Track the pulse 3d headset by Sony.  Also using "git" vs command prompt to run your code will literally tell you a debug of what codes the issue is.

**Comment by Kpa62** [comment]
The key to your error btw is this. "Property 'throttles' does not exist on type"  Throttles is missing from the Store.ts file.

**Comment by jef** [comment]
> > Do you mind posting the code you're trying to add? Specifically the new store? I can help!
> 
> Absolutely! Thanks for writing/releasing such an awesome utility, too. :-D
> 
> ```ts
> import {Store} from './store';
> 
> export const Honeycomb: Store = {
> 	labels: {
> 		inStock: {
> 			container: '#product_form_4533457682541',
> 			text: ['Add to cart']
> 		},
> 	},
> 	links: [
> 		{
> 			brand: 'test:brand',
> 			model: 'test:model',
> 			series: 'test:series',
> 			url: 'https://flyhoneycomb.com'
> 		},
> 		{
> 			brand: 'honeycomb',
> 			model: 'AlphaFlightYoke',
> 			series: 'yoke',
> 			url: 'https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/alpha-flight-controls'
> 		},
> 		{
> 			brand: 'honeycomb',
> 			model: 'BravoThrottleQuadrant',
> 			series: 'quadrant',
> 			url: 'https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/bravo-throttle-quadrant'
> 		},
> 		{
> 			brand: 'honeycomb',
> 			model: 'ThrottlePackForAirbus',
> 			series: 'throttles',
> 			url: 'https://flyhoneycomb.com/collections/honeycomb-flight-sim-hardware/products/airbus-throttle-handles'
> 		},
> 	],
> 	name: 'honeycomb'
> };
> ```

You're very close. You need to update `src/store/model/index.ts` and `src/store/model/store.ts`.

Here is the write up I made: https://jef.codes/streetmerchant/help/general/#adding-a-store

**Comment by attaway** [comment]
> I drove myself crazy over this one for a longtime but figured out how to debug. I believe there is some confusion here. I'm not a coder at all but from fiddling around I ran into this. This current error is not a store error but a product error. You need to add the model, series and brand to the store.TS file. Then you will most likely get a max price error next, so then you have to add to the config.ts the max price series. also update the .env file for this series and model in the max price section. Also update the Readme.md file. finally it should all run.
> 
> What confusion I see here though is that you're adding a store before adding a model or series for the product. you need to build both. A store and a product as the flight stick is not a tracked product by default.
> 
> I ran into this trying to Track the pulse 3d headset by Sony. Also using "git" vs command prompt to run your code will literally tell you a debug of what codes the issue is.

Wow totally solved my problem!
Just needed to edit config.ts and add the max price to .env file.
Thanks so much

@jef should take a look at this

**Comment by jef** [comment]
> @jef should take a look at this

https://jef.codes/streetmerchant/help/general/#adding-a-store

**Comment by rmp5s** [comment, marked as answer]
Wrote a whole big reply to @Kpa62 (see below), then I saw @attaway's reply.  Like @attaway, I got it working.  

For anyone that may come across this in the future, in addition to what's below and what's in @jef's write-up, I went in and added the following...

To `streetmerchant/src/config.ts`:

```
			yoke: envOrNumber(process.env.MAX_PRICE_SERIES_YOKE),
			quadrant: envOrNumber (process.env.MAX_PRICE_SERIES_QUADRANT),
			throttles: envOrNumber (process.env.MAX_PRICE_SERIES_THROTTLES),
```

To `streetmerchant/.env`:

```
MAX_PRICE_SERIES_YOKE=""
MAX_PRICE_SERIES_QUADRANT=""
MAX_PRICE_SERIES_THROTTLES=""
```

I knew the script was looking for something, I just didn't know what.  THAT is what it was looking for.

Now, whether this actually works for this website, these items, etc or not is yet to be seen, but that's on me.  

@jef - Could we get your writeup edited/ammended the above "max price" changes to the .env and config.ts files as I do not believe these changes are mentioned but seem to be required for the script to function.

Thank you all so much, and again, thanks @jef for writing/releasing this utility!  :-D 

My previous post will be in next comment as I can't get it to look right here...

**Comment by rmp5s** [comment]
Previous post - 

> This current error is not a store error but a product error. You need to add the model, series and brand to the store.TS file.

Hmm....yup.  Did that already.  Don't think so, but maybe I did it wrong?  (I commented it out so the script would still work while I got the new stuff sorted out.)

```
import {Browser, LoadEvent} from 'puppeteer';

export type Element = {
	container?: string;
	text: string[];
};

export type Pricing = {
	container: string;
	euroFormat?: boolean;
};

export type Brand =
	| 'test:brand'
	| 'amd'
	| 'asrock'
	| 'asus'
	| 'corsair'
	| 'evga'
	| 'gainward'
	| 'gigabyte'
	| 'inno3d'
	| 'kfa2'
	| 'microsoft'
	| 'msi'
	| 'nvidia'
	| 'palit'
	| 'pny'
	| 'sapphire'
	| 'sony'
	| 'xfx'
	| 'powercolor'
	| 'zotac'
/* 	| 'honeycomb' */;

export type Series =
	| 'test:series'
	| '3070'
	| '3080'
	| '3090'
	| 'rx6800'
	| 'rx6800xt'
	| 'rx6900xt'
	| 'ryzen5600'
	| 'ryzen5800'
	| 'ryzen5900'
	| 'ryzen5950'
	| 'sonyps5c'
	| 'sonyps5de'
	| 'sf'
	| 'xboxsx'
	| 'xboxss'
/* 	| 'yoke'
	| 'quadrant'
	| 'throttles' */;

export type Model =
	| 'test:model'
	| '5600x'
	| '5800x'
	| '5900x'
	| '5950x'
	| '600 platinum'
	| '750 platinum'
	| 'amd reference'
	| 'amp extreme holo'
	| 'amp holo'
	| 'aorus master'
	| 'aorus xtreme'
	| 'aorus'
	| 'challenger'
	| 'dual fan'
	| 'dual oc'
	| 'dual'
	| 'eagle oc'
	| 'eagle'
	| 'founders edition'
	| 'ftw3 ultra'
	| 'ftw3'
	| 'gamerock oc'
	| 'gaming oc'
	| 'gaming pro oc'
	| 'gaming pro'
	| 'gaming x trio'
	| 'gaming x3'
	| 'gaming'
	| 'ichill x2'
	| 'ichill x3'
	| 'ichill x4'
	| 'nitro+'
	| 'nitro oc se'
	| 'nitro oc'
	| 'phantom gaming'
	| 'phantom gs'
	| 'phoenix gs oc'
	| 'phoenix gs'
	| 'phoenix'
	| 'ps5 console'
	| 'ps5 digital'
	| 'red devil'
	| 'sg oc'
	| 'sg'
	| 'merc'
	| 'strix lc'
	| 'strix oc'
	| 'strix'
	| 'taichi'
	| 'trinity oc'
	| 'trinity'
	| 'tuf oc'
	| 'tuf'
	| 'turbo'
	| 'twin edge oc'
	| 'twin edge'
	| 'twin x2 oc'
	| 'twin x2'
	| 'uprising'
	| 'ventus 2x oc'
	| 'ventus 2x'
	| 'ventus 3x oc'
	| 'ventus 3x'
	| 'vision oc'
	| 'vision'
	| 'xbox series s'
	| 'xbox series x'
	| 'xc3 black'
	| 'xc3 ultra'
	| 'xc3'
	| 'xlr8 revel'
	| 'xlr8 uprising'
/* 	| 'AlphaFlightYoke'
	| 'BravoThrottleQuadrant'
	| 'ThrottlePackForAirbus' */;

export type Link = {
	brand: Brand;
	itemNumber?: string;
	series: Series;
	model: Model;
	url: string;
	cartUrl?: string;
	openCartAction?: (browser: Browser) => Promise<string>;
	screenshot?: string;
};

export type LabelQuery = Element[] | Element | string[];

export type Labels = {
	bannedSeller?: LabelQuery;
	captcha?: LabelQuery;
	container?: string;
	inStock?: LabelQuery;
	outOfStock?: LabelQuery;
	maxPrice?: Pricing;
};

export type StatusCodeRangeArray = Array<number | [number, number]>;

export type Store = {
	realTimeInventoryLookup?: (itemNumber: string) => Promise<boolean>;
	/**
	 * The range of status codes which will trigger backoff, i.e. an increasing
	 * delay between requests. Setting an empty array will disable the feature.
	 * If not defined, the default range will be used: 403.
	 */
	backoffStatusCodes?: StatusCodeRangeArray;
	disableAdBlocker?: boolean;
	links: Link[];
	linksBuilder?: {
		builder: (docElement: cheerio.Cheerio, series: Series) => Link[];
		ttl?: number;
		urls: Array<{series: Series; url: string | string[]}>;
	};
	labels: Labels;
	name: string;
	setupAction?: (browser: Browser) => void;
	/**
	 * The range of status codes which considered successful, i.e. without error
	 * allowing request parsing to continue. Setting an empty array will cause
	 * all requests to fail. If not defined, the default range will be used:
	 * 0 -> 399 inclusive.
	 */
	successStatusCodes?: StatusCodeRangeArray;
	waitUntil?: LoadEvent;
	minPageSleep?: number;
	maxPageSleep?: number;
};
```

> Then you will most likely get a max price error next...

Nope.  Never saw a price error...

> Also update the Readme.md file. finally it should all run.

Now this is new.  Where is that?  What do I have to do there?  This is the first I remember seeing "readme.md" anywhere...

> What confusion I see here though is that you're adding a store before adding a model or series for the product. you need to build both. A store and a product as the flight stick is not a tracked product by default.

Hmm...not sure I follow.  I added a store (honeycomb.ts) with the code shared with Jef, added the necessary bits into index.ts:
```
import {Evga} from './evga';
import {EvgaEu} from './evga-eu';
import {Galaxus} from './galaxus';
import {Game} from './game';
import {Gamestop} from './gamestop';
// import {Honeycomb} from './honeycomb';
import {Kabum} from './kabum';
import {Mediamarkt} from './mediamarkt';
import {MemoryExpress} from './memoryexpress';
import {MicroCenter} from './microcenter';
import {Mindfactory} from './mindfactory';`
```
...(again, commented out so the code will still run), and added all the stuff into store.ts as pasted above.  Following Jef's writeup and viewing the example given, as linked to in my initial post, I thought I checked all the boxes.  This "readme.md" thing though...this is new.

**Comment by Kpa62** [comment]
Ok lets go one at a time, The location of the readme is on the root of the street merchant folder where you edit the .env fix that part first. the other important thing I did was mark down what my series is, model and brand, as all these need to sync up across the grid.  

see what errors come after the readme file update.

As for a store, is honeycomb the store? you need to create a Store.TS file that pulls from a element on the webpage that determines if the  item is in stock of not, this is a process in itself, different from adding another item. that's what i was trying to say. you have 2 build outs needed to accomplish this. the current error however points to the readme file missing your model, series and brand.

**Comment by Kpa62** [comment]
for the store it would b laid out, Honeycomb.ts, within that file it would be the following. This is Walmart's for example.

import {Store} from './store';

export const Walmart: Store = {
	labels: {
		inStock: {
			container: '.button.spin-button.prod-ProductCTA--primary.button--primary',
			text: ['add to cart']
		},
		maxPrice: {
			container: 'span[class*="price-characteristic"]'
		}
	},
	links: [
		{
			brand: 'test:brand',
			model: 'test:model',
			series: 'test:series',
			url:
				'https://www.walmart.com/ip/Keurig-K-compact-Brewer-Black-Coffee-Maker/806217614'
		},
		{
			brand: 'sony',
			model: 'ps5 console',
			series: 'sonyps5c',
			url: 'https://www.walmart.com/ip/PlayStation5-Console/363472942'
		},
		{
			brand: 'sony',
			model: 'ps5 digital',
			series: 'sonyps5de',
			url:
				'https://www.walmart.com/ip/PlayStation5-Console/493824815'
		},
		{
			brand: 'microsoft',
			model: 'xbox series x',
			series: 'xboxsx',
			url: 'https://www.walmart.com/ip/Xbox-Series-X/443574645'
		},
		{
			brand: 'microsoft',
			model: 'xbox series s',
			series: 'xboxss',
			url: 'https://www.walmart.com/ip/Xbox-Series-S/606518560'
		},
		{
			brand: 'corsair',
			model: '750 platinum',
			series: 'sf',
			url: 'https://www.walmart.com/ip/SF750-Power-Supply/197046151'
		},
		{
			brand: 'corsair',
			model: '600 platinum',
			series: 'sf',
			url:
				'https://www.walmart.com/ip/Corsair-SF-Series-600W-80-Platinum-Power-Supply/250717047'
		},
		{
			brand: 'amd',
			model: '5900x',
			series: 'ryzen5900',
			url:
				'https://www.walmart.com/ip/AMD-Ryzen-9-5900X-12-core-24-thread-Desktop-Processor/159710953'
		}
	],
	name: 'walmart'
};


This file goes in SRC/store/ model.

**Reply by rmp5s** [reply]
> for the store it would b laid out, Honeycomb.ts, within that file it would be the following. This is Walmart's for example.
> 
> import {Store} from './store';
> 
> export const Walmart: Store = {
> labels: {
> inStock: {
> container: '.button.spin-button.prod-ProductCTA--primary.button--primary',
> text: ['add to cart']
> },
> maxPrice: {
> container: 'span[class*="price-characteristic"]'
> }
> },
> links: [
> {
> brand: 'test:brand',
> model: 'test:model',
> series: 'test:series',
> url:
> 'https://www.walmart.com/ip/Keurig-K-compact-Brewer-Black-Coffee-Maker/806217614'
> },
> {
> brand: 'sony',
> model: 'ps5 console',
> series: 'sonyps5c',
> url: 'https://www.walmart.com/ip/PlayStation5-Console/363472942'
> },
> {
> brand: 'sony',
> model: 'ps5 digital',
> series: 'sonyps5de',
> url:
> 'https://www.walmart.com/ip/PlayStation5-Console/493824815'
> },
> {
> brand: 'microsoft',
> model: 'xbox series x',
> series: 'xboxsx',
> url: 'https://www.walmart.com/ip/Xbox-Series-X/443574645'
> },
> {
> brand: 'microsoft',
> model: 'xbox series s',
> series: 'xboxss',
> url: 'https://www.walmart.com/ip/Xbox-Series-S/606518560'
> },
> {
> brand: 'corsair',
> model: '750 platinum',
> series: 'sf',
> url: 'https://www.walmart.com/ip/SF750-Power-Supply/197046151'
> },
> {
> brand: 'corsair',
> model: '600 platinum',
> series: 'sf',
> url:
> 'https://www.walmart.com/ip/Corsair-SF-Series-600W-80-Platinum-Power-Supply/250717047'
> },
> {
> brand: 'amd',
> model: '5900x',
> series: 'ryzen5900',
> url:
> 'https://www.walmart.com/ip/AMD-Ryzen-9-5900X-12-core-24-thread-Desktop-Processor/159710953'
> }
> ],
> name: 'walmart'
> };
> 
> This file goes in SRC/store/ model.

I got it working, my friend.  :-D  

https://github.com/jef/streetmerchant/discussions/1224#discussioncomment-167773

**Reply by Kpa62** [reply]
Great!

**Comment by rmp5s** [comment]
Now to get it running in a Docker container on my unRAID server.  I've never "manually installed" a container on there.  Gotta figure out how to do it.

Fun fun!  :-D

**Reply by jef** [reply]
You got it!
